controllers/trainingjob/trainingjob_controller.go

```diff
@@ -210,6 +210,10 @@ func (r *Reconciler) reconcileTrainingJob(ctx reconcileRequestContext) error {
 		if err = r.updateStatus(ctx, string(ctx.TrainingJobDescription.TrainingJobStatus), ""); err != nil {
 			return err
 		}
+	} else if ctx.TrainingJobDescription.FailureReason != nil {
```

Comment:
Are your proposing that we check whether it has failed, rather than if it has a failure reason? I'm not entirely certain that failure reason is populated only if the job has failed. For instance, if it has a failure reason in the `InProgress` state (I'm not sure why it would), I'd still like to add it to status.

Reply:
Ideally that should not happen.

Reply:
Do you think it is worth having as a fallback anyway?

Reply:
I also thought that the way this is written smells. I think it's because it duplicates the `updateStatus` functionality a few times, causing me to invoke the old [Rule of Three](https://en.wikipedia.org/wiki/Rule_of_three_%28computer_programming%29).

Can we remove duplication? Consider something like:

```go
primaryStatus := string(ctx.TrainingJobDescription.TrainingJobStatus)
secondaryStatus := string(ctx.TrainingJobDescription.SecondaryStatus)
additional := controllers.GetOrDefault(ctx.TrainingJobDescription.FailureReason, "")

if ctx.TrainingJobDescription.TrainingJobStatus == sagemaker.TrainingJobStatusStopping {
    // Clear the secondary status if we detected stopping, since SageMaker has unclear secondary statuses during this phase
    // Open ticket with the SageMaker team: https://t.corp.amazon.com/0411302791
    secondaryStatus = ""
}

if err = r.updateStatusWithAdditional(ctx, primaryStatus, secondaryStatus, additional); err != nil {
	return err
}
```

It might make sense to move the few lines of code that generate the statuses into their own function. I'll leave that up to you.

Reply:
Great suggestion! Did not know about the `GetOrDefault` method, love it.